File: app/src/androidTest/java/com/bopr/android/smailer/HtmlMatcher.java
package com.bopr.android.smailer;

import org.hamcrest.CustomTypeSafeMatcher;
import org.hamcrest.Description;
import org.hamcrest.Matcher;

import java.io.IOException;
import java.io.InputStream;
import java.util.Scanner;
import java.util.regex.Pattern;

import static androidx.test.platform.app.InstrumentationRegistry.getInstrumentation;
import static com.bopr.android.smailer.util.Util.readStream;

class HtmlMatcher extends CustomTypeSafeMatcher<String> {

    private final String expected;
    private String expectedToken;
    private String actualToken;

    static Matcher<String> htmlEquals(String expected) {
        return new HtmlMatcher(expected);
    }

    static Matcher<String> htmlEqualsRes(String resource) {
        try {
            InputStream stream = getInstrumentation().getContext().getAssets().open(resource);
            String expected = readStream(stream);
            stream.close();
            return htmlEquals(expected);
        } catch (IOException x) {
            throw new IllegalArgumentException(x);
        }
    }


    private HtmlMatcher(String expected) {
        super(expected);
        this.expected = expected;
    }

    @Override
    protected boolean matchesSafely(String actual) {
        Pattern delimiters = Pattern.compile("(\\s|>|<|;)+");
        Scanner es = new Scanner(expected).useDelimiter(delimiters);
        Scanner as = new Scanner(actual).useDelimiter(delimiters);

        expectedToken = null;
        actualToken = null;

        while (es.hasNext() && as.hasNext()) {
            expectedToken = es.next();
            actualToken = as.next();
            if (!expectedToken.equals(actualToken)) {
                return false;
            }
        }
        return es.hasNext() == as.hasNext();
    }

    @Override
    protected void describeMismatchSafely(String item, Description description) {
        super.describeMismatchSafely(item, description);
        description
                .appendText("\nExpected token: '")
                .appendText(expectedToken)
                .appendText("' but found: '")
                .appendText(actualToken)
                .appendText("'");
    }
}

File: app/src/androidTest/kotlin/com/bopr/android/smailer/HtmlMatcher.kt
package com.bopr.android.smailer

import androidx.test.platform.app.InstrumentationRegistry
import com.bopr.android.smailer.util.readStream
import org.hamcrest.CustomTypeSafeMatcher
import org.hamcrest.Description
import java.io.IOException
import java.util.*
import java.util.regex.Pattern

/**
 * Utility class to match HTML files ignoring whitespaces.
 */
internal open class HtmlMatcher private constructor(private val expected: String) :
        CustomTypeSafeMatcher<String?>("HTML matches") {

    private lateinit var actual: String
    private var expectedToken: String? = null
    private var actualToken: String? = null
    private val delimiters = Pattern.compile("(\\s|>|<|;)+")

    override fun matchesSafely(html: String?): Boolean {
        this.actual = html!!
        if (actual == expected) {
            return true
        } else {
            val expects = Scanner(expected).useDelimiter(delimiters)
            val actuals = Scanner(actual).useDelimiter(delimiters)
            expectedToken = null
            actualToken = null
            while (expects.hasNext() && actuals.hasNext()) {
                expectedToken = expects.next()
                actualToken = actuals.next()
                if (expectedToken != actualToken) {
                    return false
                }
            }
            return expects.hasNext() == actuals.hasNext()
        }
    }

    override fun describeMismatchSafely(item: String?, description: Description) {
        description
                .appendText("Expected token [")
                .appendText(expectedToken)
                .appendText("] but [")
                .appendText(actualToken)
                .appendText("] found.")

        description
                .appendText("\n-------\n")
                .appendText(actual)
    }

    companion object {

        fun htmlEquals(resourceName: String): HtmlMatcher {
            try {
                InstrumentationRegistry.getInstrumentation().context.assets.open(resourceName).use {
                    return HtmlMatcher(readStream(it))
                }
            } catch (x: IOException) {
                throw IllegalArgumentException("Invalid resource", x)
            }
        }
    }


}